File: 045 - Triangular, pentagonal, and hexagonal/C++/pe_045.cc
/*
 * file:    pe_045.cc
 * title:   Triangular, pentagonal, hexagonal
 * author:  Alexander P. Gosselin
 * e-mail:  alexandergosselin@gmail.com
 * date:    September 23, 2014
 * 
 * math:    Triangular, pentagonal, and hexagonal numbers have the forms
 *          n(n + 1)/2, n(3n - 1)/2, and n(2n - 1) respectively.
 *          Every hexagonal number is also a triangular number, since
 *          n(2n - 1) = (2n-1)[(2n - 1) + 1]/2
 * 
 *          http://mathworld.wolfram.com/HexagonalNumber.html
 */

#include <cmath>
#include <iostream>

bool is_pentagonal(size_t);

const int min_hexagonal_index = 144;

using namespace std;
int main() {
  for (size_t n = min_hexagonal_index; true; n++) {
    size_t hexagonal = n*(2*n - 1);
    if (is_pentagonal(hexagonal)) {
      cout << hexagonal << endl;
      break;
    }
  }
}

bool is_pentagonal(size_t n) {
  double x = (1 + sqrt(1 + 24*n))/6.0;
  if (floor(x) == x) 
    return true;
  else 
    return false;
}

File: 045 - Triangular, pentagonal, and hexagonal/Haskell/pe_045.hs
{-
file:   pe_045.hs
title:  Triangular, pentagonal, and hexagonal
author: Alexander P. Gosselin
e-mail: alexandergosselin@gmail.com
date:   May 25, 2015

link:   https://projecteuler.net/problem=45

copyright:  (C) 2015 Alexander Gosselin
license:    GNU General Public License <http://www.gnu.org/licenses/>
-}

triangular :: [Integer]
triangular = map (\x -> x*(x + 1) `quot` 2)   [1..]

pentagonal :: [Integer]
pentagonal = map (\x -> x*(3*x - 1) `quot` 2) [1..]

hexagonal  :: [Integer]
hexagonal  = map (\x -> x*(2*x - 1))          [1..]

intersection :: Ord a => [a] -> [a] -> [a] -> [a]
intersection []     (y:ys) (z:zs) = []
intersection (x:xs) []     (z:zs) = []
intersection (x:xs) (y:ys) []     = []
intersection (x:xs) (y:ys) (z:zs)
    | x < y     = intersection xs (y:ys) (z:zs)
    | y < z     = intersection (x:xs) ys (z:zs)
    | z < x     = intersection (x:xs) (y:ys) zs
    | otherwise = x:intersection xs ys zs -- x == y == z

pe_045 = head 
    $ filter ((<) 40755) 
    $ intersection triangular pentagonal hexagonal

main :: IO ()
main = do
    print pe_045
